src/tun2socks/build.action.mjs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2023 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import url from 'url';
+import os from 'node:os';
+import path from 'node:path';
+import fs from 'node:fs/promises';
+
+import {spawnStream} from '../build/spawn_stream.mjs';
+import {getBuildParameters} from '../build/get_build_parameters.mjs';
+
+/**
+ * @description Builds the tun2socks library for the specified platform.
+ *
+ * @param {string[]} parameters
+ */
+export async function main(...parameters) {
+  const {platform: targetPlatform} = getBuildParameters(parameters);
+
+  const binDir = path.join(process.env.OUTPUT_DIR, 'bin');
+  const buildDir = path.join(
+    process.env.BUILD_DIR,
+    ['ios', 'macos', 'maccatalyst'].includes(targetPlatform) ? 'apple' : targetPlatform
+  );
+
+  await fs.mkdir(binDir, {recursive: true});
+  await fs.mkdir(buildDir, {recursive: true});
+
+  // install go tools locally
+  await spawnStream(
+    'go',
+    'build',
+    '-o',
+    binDir,
+    'golang.org/x/mobile/cmd/gomobile',
+    'golang.org/x/mobile/cmd/gobind',
+    'github.com/crazy-max/xgo'
+  );
+
+  const hostPlatform = os.platform();
+
+  process.env.PATH = `${binDir}${path.delimiter}${process.env.PATH}`;
+
+  switch (targetPlatform + hostPlatform) {
+    case 'android' + 'darwin':
+    case 'android' + 'linux':
+    case 'android' + 'win32':
+      return spawnStream(
+        'gomobile',
+        'bind',
+        '-androidapi=33',
+        '-target=android',
+        `-o=${buildDir}/tun2socks.aar`,
+
+        'github.com/Jigsaw-Code/outline-client/src/tun2socks/outline/tun2socks',
+        'github.com/Jigsaw-Code/outline-client/src/tun2socks/outline/shadowsocks'
+      );
+    case 'ios' + 'darwin':
+    case 'macos' + 'darwin':
+    case 'maccatalyst' + 'darwin':
+      process.env.MACOSX_DEPLOYMENT_TARGET = '10.14';
+
+      return spawnStream(
+        'gomobile',
+        'bind',
+        '-bundleid=org.outline.tun2socks',
+        '-iosversion=13.1',
```

Comment:
This is a change in behavior. We have not agreed to upgrade iOS to 13.1. I think we agreed to upgrade to iOS 12 only.

The current behavior has two different targets: one for iOS and one for macOS, including Catalyst, then a [step that merges them](https://github.com/Jigsaw-Code/outline-go-tun2socks/blob/37e94192a607f78023b3191b0b909fd6cf9f874e/Makefile#L40).
That way we can specify different iOS versions, since Catalyst requires 13.1.

Reply:
That could work, but the iOS/macCatalyst project is shared and refers to the same file location. So you will need to use the same path for both, or change the project to point to different files based on the target.

I found it very useful for development to be able to build and run both the iOS and the macCatalyst apps, without having to rebuild the project, and I'd like to keep that behavior because it lets us run the VPN, without needing a phone.

/cc @sbruens for thoughts

Reply:
For now, can we restore how this was done?  I mean, two gomobile calls with a merge to build one framework for all platforms. This needs to be incremental though, if it's running on every build, since the framework build takes some time.

Reply:
The old behavior set different iOS versions for iOS and macCatalyst, then merged the XCFrameworks. That is how Sander and I figured out how to support different versions across the platforms. We explicitly wanted that behavior because of the better development experience and simplicity of use. It's also helpful to know that a file is always the same thing. If it changes with the platform, perhaps it should have a different name or path (like we do for the cordova projects), but then the XCode project needs to be tweaked too.

The other issue is the incremental build. Perhaps it changed, but my past experience is that gomobile discarded a lot of the work done between runs.

I believe the easiest is to preserve the Makefile we had and just call make in the build action. It minimizes change, preserves the merged XCFramework and the incremental build we had before.